docs/design/caching/README.md

```diff
@@ -0,0 +1,398 @@
+    end
+
+    participant E as [Transactional DB]
+
+    A->>B: upsert(request)
+    B->>C: upsertDocument(request)
+    C->>D: upsertDocument(request, client)
+
+    D->>F: validateReferences(...)
+    F->>E: "select" statement
+    E-->>F: result
+    F-->>D: result
+
+    opt result == all exist
+      D->>E: upsert(doc)
+      D-->>C: success
+    end
+    alt
+      D-->>C: failed
+    end
+
+    C-->>B: response
+    B-->>A: response
+```
+
+> **Note**
+> In subsequent sequence diagrams, the `FrontEndFacade` will be
+> omitted.
+
+### Introducing a Cache Layer
+
+One simple pattern is to introduce an [Identity
+Map](https://martinfowler.com/eaaCatalog/identityMap.html), which could be
+implemented with multiple backend providers, representing in-memory or external
+caches. It will be convenient to split this module in two, so that common
+business logic remains in the initial Identity Map module, which then delegates
+out to a simple provider that encapsulates data access.
+
+Typically the Identity Map module would sit between the backend facade (as
+labeled above) and the repository layer. However, we have validation logic
+inside the repository so that the validation checks are inside a transaction
+context. Therefore the repository layer will need to know how to access the
+cache layer.
+
+To prevent the repository layer from accruing too much cache-related code, cache
+management needs to be centralized in a `Cache` module (the Identity Map), which
+can be placed in the `meadowlark-core` package. This also aids in avoid coding
+duplication between different backend implementations. There will need to be
+startup code to inject the right provider into the `Cache` module; these two are
+treated as one "cache" in the sequence diagram below for simplicity. Then the
+core `Upsert` module can inject the cache into the backend calls following the
+Dependency Inversion principle.
+
+The repository layer will ask the cache if the references exist. If the cache
+does not contain the references, then it needs to go back to the repository
+layer to check the transactional database. Any hits should be added to the cache
+at that point, for use by subsequent API requests that need to validate the same
+references. For efficiency when querying external provides, the Cache interface
+needs to receive an array of ID's to lookup, rather than just a single ID each
+time. Thus the code can use batch requests instead of individual service
+requests for each ID.
+
+```mermaid
+sequenceDiagram
+    box Core
+        participant B as Upsert
+        participant G as Cache
+    end
+
+    box Backend
+        participant C as BackendFacade
+        participant D as Backend.Upsert
+        participant F as Backend.ReferenceValidation
+    end
+
+    participant E as [Transactional DB]
+
+    B->>C: upsertDocument(request, cache)
+    C->>D: upsertDocument(request, cache, client)
+    D->>D: startTransaction()
+
+    rect ivory
```

Comment:
There are a few rectangles that have a specific color (I guess that's from the *ivory* definition here) that do not look correctly on dark themes from GitHub, we should use another color to avoid this cases. In my case, the dark theme is my default even as incognito an makes it difficult to understand what's happening.

![image](https://github.com/Ed-Fi-Exchange-OSS/Meadowlark/assets/5459528/98d65bd7-025b-4aca-9c38-20bac4af2972)

Reply:
Yes, that is hard to read. I don't know how to change the text color. The ivory was used to highlight what has changed. Maybe I can find a way to make the entire diagram have a white background even when in a dark theme.